client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java

```diff
@@ -907,8 +908,43 @@ public final void shutdown() {
     @Override
     public final void shutdown(final Handler<AsyncResult<Void>> completionHandler) {
 
-        if (shuttingDown.compareAndSet(Boolean.FALSE, Boolean.TRUE)) {
+        closeConnection(completionHandler, true);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public final void disconnect() {
+
+        disconnect(disconnectResult -> {
+            if (disconnectResult.succeeded()) {
+                LOG.info("successfully disconnected from the server");
+            } else {
+                LOG.error("could not disconnect from the server", disconnectResult.cause());
+            }
+        });
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public final void disconnect(final Handler<AsyncResult<Void>> completionHandler) {
+        closeConnection(completionHandler, false);
+    }
+
+    //-----------------------------------< private methods >---
+
+    private void closeConnection(final Handler<AsyncResult<Void>> completionHandler, final Boolean isShuttingDown) {
+
+        Objects.requireNonNull(completionHandler, "Result handler cannot be null");
+        AtomicBoolean flag = (isShuttingDown) ? shuttingDown : disconnecting;
```

Comment:
I think, deriving only one flag from `shuttingDown` and `disconnecting`, and using that flag for the subsequent check, could cause issues here and makes the code harder to understand.
For example the line below with the "already shutting down" exception doesn't fit if `isShuttingDown` is false and `disconnect()` is invoked concurrently.
On the other hand, calling `shutdown()` and subsequently having `disconnect()` be called, could trigger `shutdownConnection()` twice.

And I would rename the `isShuttingDown` method parameter (e.g. to `isShuttingDownParam`) to have it be more distinguishable from the `shuttingDown` field.

In general, having the `closeConnection()` method change the `shuttingDown` field could be considered an unexpected sideeffect of the method. So, I think it would be better to change the field inside the `shutdown()` method,

Reply:
thanks for the comments! i will update the PR